File: day2/sol.py
with open('input.txt', 'r') as file:
    rounds = file.readlines()

shape_reward = {
    'X': 1,
    'Y': 2,
    'Z': 3,
}

# A - Rock
# B - Paper
# C - Scissors
# X - Rock
# Y - Paper
# Z - Scissors

outcome_reward = {
    'AX': 3,  # rock rock
    'AY': 6,  # rock paper
    'AZ': 0,  # rock scissors
    'BX': 0,  # paper rock
    'BY': 3,  # paper paper
    'BZ': 6,  # paper scissors
    'CX': 6,  # scissors rock
    'CY': 0,  # scissors paper
    'CZ': 3,  # scissors scissors
}

# X - Loose
# Y - Draw
# Z - Win

# rock 1
# paper 2
# scissors 3

outcome_reward_2 = {
    'X': 0,
    'Y': 3,
    'Z': 6
}

shape_reward_2 = {
    'AX': 3,  # rock scissors
    'AY': 1,  # rock rock
    'AZ': 2,  # rock paper
    'BX': 1,  # paper rock
    'BY': 2,  # paper paper
    'BZ': 3,  # paper scissors
    'CX': 2,  # scissors paper
    'CY': 3,  # scissors scissors
    'CZ': 1,  # scissors rock
}

score = 0
for round in rounds:
    round = round.rstrip()
    round = ''.join(round.split(' '))
    score += shape_reward[round[-1]] + outcome_reward[round]

print(f'Round 1 Final score - {score}')

score2 = 0
for round in rounds:
    round = round.rstrip()
    round = ''.join(round.split(' '))
    score2 += shape_reward_2[round] + outcome_reward_2[round[-1]]
print(f'Round 2 Final score {score2}')

File: day2/sol.hs
(part1, part2) = (sum . fmap p1 <$> inp, sum . fmap p2 <$> inp)
  where
    inp = fmap words . lines <$> readFile "day2/input.txt"
    p1 xs = case (head xs, xs !! 1) of
      ("A", "X") -> 4 -- Rock, Rock
      ("B", "Y") -> 5 -- Paper, Paper
      ("C", "Z") -> 6 -- Scissors, Scissors
      ("B", "X") -> 1 -- Paper, Rock
      ("A", "Y") -> 8 -- Rock, Paper
      ("A", "Z") -> 3 -- Rock, Scissors
      ("B", "Z") -> 9 -- Paper, Scissors
      ("C", "X") -> 7 -- Scissors, Rock
      ("C", "Y") -> 2 -- Scissors, Paper
    p2 xs = case (head xs, xs !! 1) of
      ("A", "X") -> 3 -- Rock, Lose -> Scissors
      ("B", "X") -> 1 -- Paper, Lose -> Rock
      ("C", "X") -> 2 -- Scissors, Lose -> Paper
      ("A", "Y") -> 4 -- Rock, Tie
      ("B", "Y") -> 5 -- Paper, Tie
      ("C", "Y") -> 6 -- Scissors, Tie
      ("A", "Z") -> 8 -- Rock, Win  -> Paper
      ("B", "Z") -> 9 -- Paper, Win -> Scissors
      ("C", "Z") -> 7 -- Scissors, Win -> Rock
